### PR TITLE
v4/rest api

### PR DESCRIPTION
**Pull request by alexandrebodin**
What does it do ? 

This PR Implements the Core API changes for V4.

Amond those we have: `populate` / `fields` / `sort` and `pagination` new syntax. 

- Cleanup parse multipart
- Transform response relations to use data & meta
- Suport media format
- Init populate parsing
- Implement rest api populate syntax and Init  refactor convert query params for v4
- Support booleans for nested populates
- Implement fields convert query params
- Handle sort query param
- Implement REST API pagination with paged or offset and withCount
- Implement pagination logic in core api
- Fix tests
- Refactor request format in core-api

### CODE REVIEW

File: packages/core/strapi/lib/core-api/service/pagination.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  if (maxLimit && (limit === -1 || limit > maxLimit)) {
+    return maxLimit;
+  }
+
+  return limit;
+};
+
+const shouldCount = params => {
+  if (_.has(params, 'pagination.withCount')) {
+    return params.pagination.withCount === 'false' ? false : true;
```
Comment (petersg83) [inline packages/core/strapi/lib/core-api/service/pagination.js:31]: Can the param be `false`? In the case of a dev extend the service for instance of else. If yes, maybe we should handle the case of the boolean `false`.

File: packages/core/strapi/lib/core-api/service/pagination.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  return Boolean(strapi.config.get('api.rest.withCount', true));
+};
+
+const isOffsetPagination = pagination => has('start', pagination) || has('limit', pagination);
+const isPagedPagination = pagination => has('page', pagination) || has('page', pagination);
```
Comment (petersg83) [inline packages/core/strapi/lib/core-api/service/pagination.js:38]: is the second one supposed to be `pageSize`?

File: packages/core/strapi/lib/core-api/service/pagination.js (hunk 1 above)
Comment (alexandrebodin) [reply packages/core/strapi/lib/core-api/service/pagination.js:31]: I don't understand do you mean test for a boolean type value ?

Comment (petersg83) [reply packages/core/strapi/lib/core-api/service/pagination.js:31]: yes exactly

File: packages/core/utils/lib/convert-query-params.js [hunk 3]
```diff
@@ -0,0 +1,236 @@
+'use strict';
+
+/**
+ * Converts the standard Strapi REST query params to a moe usable format for querying
```
Comment (petersg83) [inline packages/core/utils/lib/convert-query-params.js:4]: more*

File: packages/core/utils/lib/convert-query-params.js [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+ * You can read more here: https://strapi.io/documentation/developer-docs/latest/developer-resources/content-api/content-api.html#filters
+ */
+
+const _ = require('lodash');
+
+// const BOOLEAN_OPERATORS = ['or', 'and'];
```
Comment (petersg83) [inline packages/core/utils/lib/convert-query-params.js:10]: to be removed?

File: packages/core/utils/lib/convert-query-params.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+  constructor() {
+    super();
+    this.message = 'Invalid order. order can only be one of asc|desc|ASC|DESC';
+  }
+}
+class InvalidSortError extends Error {
+  constructor() {
+    super();
+    this.message =
+      'Invalid sort parameter. Expected a string, an array of strings, a sort object or an array of sort object';
```
Comment (petersg83) [inline packages/core/utils/lib/convert-query-params.js:23]: an array of sort object**s***?

File: packages/core/utils/lib/convert-query-params.js [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  if (!['asc', 'desc'].includes(order.toLocaleLowerCase())) {
+    throw new InvalidOrderError();
+  }
+};
+
+/**
+ * Sort query parser
+ * @param {string} sortQuery - ex: id:asc,price:desc
+ */
+const convertSortQueryParams = sortQuery => {
```
Comment (petersg83) [inline packages/core/utils/lib/convert-query-params.js:37]: The return value is not always the same format ? sometimes an array and sometimes an object ?

File: packages/core/utils/lib/convert-query-params.js [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+      transformedSort[field] = order;
+    }
+  }
+
+  return transformedSort;
+};
+
+/**
+ * Start query parser
+ * @param {string} startQuery - ex: id:asc,price:desc
```
Comment (petersg83) [inline packages/core/utils/lib/convert-query-params.js:85]: bad copy paste I think

File: packages/core/utils/lib/convert-query-params.js [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+  if (!_.isInteger(startAsANumber) || startAsANumber < 0) {
+    throw new Error(`convertStartQueryParams expected a positive integer got ${startAsANumber}`);
+  }
+
+  return startAsANumber;
+};
+
+/**
+ * Limit query parser
+ * @param {string} limitQuery - ex: id:asc,price:desc
```
Comment (petersg83) [inline packages/core/utils/lib/convert-query-params.js:99]: bad copy paste I think

File: packages/core/utils/lib/index.js [hunk 9]
```diff
@@ -4,11 +4,7 @@
  * Export shared utilities
  */
 const { buildQuery, hasDeepFilters } = require('./build-query');
-const {
-  convertRestQueryParams,
```
Comment (petersg83) [inline packages/core/utils/lib/index.js:8]: Why is it not exported anymore? The id is to migrate the code directly in Strapi package?

File: packages/core/strapi/tests/search.test.e2e.js [hunk 10]
```diff
@@ -330,14 +287,6 @@ describe('Search query', () => {
...
       expect(res.body.data.length).toBe(1);
       expect(res.body.data[0]).toMatchObject(transformToRESTResource(data.bed[0]));
-      expect(res.body.meta).toMatchObject({
-        pagination: {
-          page: 1,
-          pageSize: 10,
-          pageCount: 1,
-          total: 1,
-        },
-      });
```
Comment (petersg83) [inline packages/core/strapi/tests/search.test.e2e.js:340]: Why did you delete this part?

File: packages/core/utils/lib/convert-query-params.js (hunk 6 above)
Comment (alexandrebodin) [reply packages/core/utils/lib/convert-query-params.js:37]: Yes the db layer can handle multiple formats

File: packages/core/strapi/tests/search.test.e2e.js (hunk 10 above)
Comment (alexandrebodin) [reply packages/core/strapi/tests/search.test.e2e.js:340]: I added it by mistake when I was fixing the search test. it didn't make much sense to check the pagination on the search test we should have dedicated tests :)

File: packages/core/utils/lib/index.js (hunk 9 above)
Comment (alexandrebodin) [reply packages/core/utils/lib/index.js:8]: Just because I deleted the function but I'll refactor the code again once the content api is a bit more ready

File: packages/core/strapi/lib/core-api/service/pagination.js (hunk 1 above)
Comment (alexandrebodin) [reply packages/core/strapi/lib/core-api/service/pagination.js:31]: SUre I can add it

**Review by petersg83 — APPROVED** [review]
LGTM :)
